File: backend/src/agents/community.ts
import fs from 'node:fs';
import path from 'node:path';
import { randomUUID } from 'node:crypto';
import { db } from '../db';
import { Group, PrayerRequest, Discussion, DiscussionReply } from '../types';
import { moderate } from './communityModeration';
import { PrayerRepository, DiscussionRepository } from '../repositories';

let seedGroupsCache: Group[] | null = null;
function getSeedGroups(): Group[] {
  if (!seedGroupsCache) {
    seedGroupsCache = JSON.parse(
      fs.readFileSync(path.join(__dirname, '..', 'data', 'groups.json'), 'utf-8')
    ).groups.map((g: any) => ({ ...g, memberCount: 0 }));
  }
  return seedGroupsCache!;
}

function ensureGroupsSeeded() {
  const database = db.read();
  if (database.groups.length === 0) {
    database.groups = getSeedGroups();
    db.write(database);
  }
}

export function listGroups() {
  ensureGroupsSeeded();
  const database = db.read();
  return database.groups.map(g => ({
    ...g,
    memberCount: database.groupMembers.filter(m => m.groupId === g.id).length,
  }));
}

export function joinGroup(groupId: string, deviceId: string) {
  ensureGroupsSeeded();
  const database = db.read();
  const group = database.groups.find(g => g.id === groupId);
  if (!group) throw new Error('Group not found');

  const already = database.groupMembers.some(m => m.groupId === groupId && m.deviceId === deviceId);
  if (!already) {
    database.groupMembers.push({ groupId, deviceId, joinedAt: new Date().toISOString() });
    db.write(database);
  }
  return { joined: true, memberCount: database.groupMembers.filter(m => m.groupId === groupId).length };
}

export async function listPrayerRequests() {
  return PrayerRepository.listApproved();
}

export async function submitPrayerRequest(deviceId: string, text: string, category?: string): Promise<{ request: PrayerRequest; needsSupportNote: boolean }> {
  const { status, needsSupportNote } = moderate(text);
  const request: PrayerRequest = {
    id: randomUUID(),
    deviceId,
    authorName: 'A LifeBook user',
    text,
    category,
    prayerCount: 0,
    moderationStatus: status,
    createdAt: new Date().toISOString(),
  };

  await PrayerRepository.create({
    id: request.id,
    userId: deviceId,
    text,
    category,
    moderationStatus: status,
    aiFlaggedReason: needsSupportNote ? 'Distress or escalation keywords flagged' : undefined,
  });

  return { request, needsSupportNote };
}

export async function prayFor(requestId: string) {
  const newCount = await PrayerRepository.pray(requestId);
  const local = db.read();
  const req = local.prayerRequests.find(r => r.id === requestId) || {
    id: requestId,
    deviceId: '',
    authorName: 'A LifeBook user',
    text: '',
    prayerCount: newCount,
    moderationStatus: 'approved' as const,
    createdAt: new Date().toISOString(),
  };
  req.prayerCount = newCount;
  return req;
}

export async function listDiscussions() {
  return DiscussionRepository.listApproved();
}

export async function createDiscussion(deviceId: string, title: string, body: string, tags: string[] = []): Promise<{ discussion: Discussion; needsSupportNote: boolean }> {
  const { status, needsSupportNote } = moderate(`${title}\n${body}`);
  const discussion: Discussion = {
    id: randomUUID(),
    deviceId,
    authorName: 'A LifeBook user',
    title,
    body,
    tags,
    replyCount: 0,
    likeCount: 0,
    moderationStatus: status,
    createdAt: new Date().toISOString(),
  };

  await DiscussionRepository.create({
    id: discussion.id,
    userId: deviceId,
    title,
    body,
    tags,
    moderationStatus: status,
    aiFlaggedReason: needsSupportNote ? 'Flagged by community standards' : undefined,
  });

  return { discussion, needsSupportNote };
}

export async function likeDiscussion(discussionId: string) {
  return DiscussionRepository.like(discussionId);
}

export async function listReplies(discussionId: string) {
  return DiscussionRepository.listReplies(discussionId);
}

export async function replyToDiscussion(discussionId: string, deviceId: string, text: string): Promise<{ reply: DiscussionReply; needsSupportNote: boolean }> {
  const { needsSupportNote } = moderate(text);
  const id = randomUUID();
  const reply = await DiscussionRepository.reply(discussionId, {
    id,
    userId: deviceId,
    text,
  });
  return { reply, needsSupportNote };
}

File: backend/app/agents/community.py
import uuid
from datetime import datetime, timezone
from typing import List, Dict, Any, Optional
from ..db import get_db
from ..models.schemas import Group, PrayerRequest, Discussion, DiscussionReply

def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()

def moderate_text(text: str) -> Dict[str, Any]:
    CRISIS_KEYWORDS = ["suicide", "kill myself", "want to die", "end my life", "self harm"]
    EXPLICIT_KEYWORDS = ["hate", "violence", "kill", "attack"]
    
    haystack = text.lower()
    needs_support = any(k in haystack for k in CRISIS_KEYWORDS)
    needs_review = any(k in haystack for k in EXPLICIT_KEYWORDS) or needs_support
    
    return {
        "status": "pending" if needs_review else "approved",
        "needsSupportNote": needs_support
    }

def list_groups() -> List[Dict[str, Any]]:
    db = get_db().read()
    groups = db.get("groups", [])
    members = db.get("groupMembers", [])
    
    result = []
    for g in groups:
        count = sum(1 for m in members if m.get("groupId") == g.get("id"))
        base_count = g.get("memberCount", 0)
        result.append({
            **g,
            "memberCount": max(base_count, count)
        })
    return result

def join_group(group_id: str, device_id: str) -> Dict[str, Any]:
    db = get_db().read()
    group = next((g for g in db.get("groups", []) if g.get("id") == group_id), None)
    if not group:
        raise ValueError("Group not found")
        
    members = db.get("groupMembers", [])
    already = any(m.get("groupId") == group_id and m.get("deviceId") == device_id for m in members)
    if not already:
        members.append({
            "groupId": group_id,
            "deviceId": device_id,
            "joinedAt": _now_iso()
        })
        db["groupMembers"] = members
        get_db().write(db)
        
    total_count = sum(1 for m in members if m.get("groupId") == group_id)
    return {
        "joined": True,
        "memberCount": max(group.get("memberCount", 0), total_count)
    }

def list_prayer_requests() -> List[Dict[str, Any]]:
    db = get_db().read()
    requests = [r for r in db.get("prayerRequests", []) if r.get("moderationStatus") == "approved"]
    requests.sort(key=lambda r: r.get("createdAt", ""), reverse=True)
    return requests

def submit_prayer_request(device_id: str, text: str, category: Optional[str] = "General") -> Dict[str, Any]:
    mod = moderate_text(text)
    req_id = str(uuid.uuid4())
    request_obj = {
        "id": req_id,
        "deviceId": device_id,
        "authorName": "A LifeBook user",
        "text": text,
        "category": category,
        "prayerCount": 0,
        "moderationStatus": mod["status"],
        "createdAt": _now_iso()
    }
    
    db = get_db().read()
    db.setdefault("prayerRequests", []).append(request_obj)
    get_db().write(db)
    
    return {
        "request": request_obj,
        "needsSupportNote": mod["needsSupportNote"]
    }

def pray_for(request_id: str) -> Dict[str, Any]:
    db = get_db().read()
    req = next((r for r in db.get("prayerRequests", []) if r.get("id") == request_id), None)
    if not req:
        raise ValueError("Prayer request not found")
    req["prayerCount"] = req.get("prayerCount", 0) + 1
    get_db().write(db)
    return req

def list_discussions() -> List[Dict[str, Any]]:
    db = get_db().read()
    discussions = [d for d in db.get("discussions", []) if d.get("moderationStatus") == "approved"]
    discussions.sort(key=lambda d: d.get("createdAt", ""), reverse=True)
    return discussions

def create_discussion(device_id: str, title: str, body: str, tags: Optional[List[str]] = None) -> Dict[str, Any]:
    mod = moderate_text(f"{title} {body}")
    disc_id = str(uuid.uuid4())
    disc_obj = {
        "id": disc_id,
        "deviceId": device_id,
        "authorName": "Believer",
        "title": title,
        "body": body,
        "tags": tags or [],
        "replyCount": 0,
        "likeCount": 0,
        "moderationStatus": mod["status"],
        "createdAt": _now_iso()
    }
    
    db = get_db().read()
    db.setdefault("discussions", []).append(disc_obj)
    get_db().write(db)
    
    return {
        "discussion": disc_obj,
        "needsSupportNote": mod["needsSupportNote"]
    }

def like_discussion(discussion_id: str) -> Dict[str, Any]:
    db = get_db().read()
    disc = next((d for d in db.get("discussions", []) if d.get("id") == discussion_id), None)
    if not disc:
        raise ValueError("Discussion not found")
    disc["likeCount"] = disc.get("likeCount", 0) + 1
    get_db().write(db)
    return disc

def list_replies(discussion_id: str) -> List[Dict[str, Any]]:
    db = get_db().read()
    replies = [r for r in db.get("discussionReplies", []) if r.get("discussionId") == discussion_id]
    replies.sort(key=lambda r: r.get("createdAt", ""))
    return replies

def reply_to_discussion(discussion_id: str, device_id: str, text: str) -> Dict[str, Any]:
    db = get_db().read()
    disc = next((d for d in db.get("discussions", []) if d.get("id") == discussion_id), None)
    if not disc:
        raise ValueError("Discussion not found")
        
    reply_id = str(uuid.uuid4())
    reply_obj = {
        "id": reply_id,
        "discussionId": discussion_id,
        "deviceId": device_id,
        "authorName": "Believer",
        "text": text,
        "createdAt": _now_iso()
    }
    
    disc["replyCount"] = disc.get("replyCount", 0) + 1
    db.setdefault("discussionReplies", []).append(reply_obj)
    get_db().write(db)
    return reply_obj
